### DISCUSSION TITLE
dockerized app just outputs Hello, world! and exits.

### DISCUSSION THREAD

**Discussion by olekspickle** [rwf2/Rocket#2517] [Questions] [answered]
I have a more or less [simple template](https://github.com/alekspickle/rocket-template) and simple `cargo run` works like a charm but when I build docker image and try to run it I get:

```
docker build -t rocket-template:local .
docker run --rm -it -p 8000:8000 --hostname test-rocket --name test-rocket rocket-template:local
Hello, world!

```
I am so confused.
I have a multistage dockerfile to leverage build time and debian slim image so that the resulting image would be as small as possible.

I read all relatable issues and discussions I could find. I don't get where the hell this hello world could be, it is definitely not in my project.
AFAIK the common issues are Rocket.toml and some proxies, but in my case I want at least run the simple server in docker locally for start, and I most certainly do COPY it to the last stage image.

Any ideas?

**Comment by SergioBenitez** [comment, marked as answer]
Here's a `Dockerfile` that I recommend using as a template:

```dockerfile
FROM docker.io/rust:1.68.2-slim-bullseye AS builder

WORKDIR /build

COPY . .
RUN --mount=type=cache,target=/build/target \
    --mount=type=cache,target=/usr/local/cargo/registry \
    --mount=type=cache,target=/usr/local/cargo/git \
    --mount=type=cache,target=/usr/local/rustup \
    set -eux; \
	rustup install stable; \
    cargo build --release; \
    objcopy --compress-debug-sections target/release/server ./server

################################################################################

FROM docker.io/debian:bullseye-slim

WORKDIR /app

COPY --from=builder /build/server ./server
CMD ROCKET_ADDRESS=0.0.0.0 ROCKET_PORT=$PORT ./server
```

Build and run:

```sh
docker build -t server:local .
docker run -e=PORT=8000 -dt -p 8000:8000/tcp --rm server:local
```

Can you give this a try and see if there's an issue?

**Reply by olekspickle** [reply]
Well, it takes initially hell of a lot more time that taking musl at the first place, but I guess it's just a tradeoff time vs space: having a heavier cache image vs installing rust in a slimmer one.
After cached image ready it is pretty fast and works as expected so I guess it's better.

Also cache `/build` name messes up fileserver since it is apparently included at compile time so the current directory in cache and resulting image should share a WORKDIR name:
```
FileServer path '/build/static/assets' is not a directory.
```
but that's fine for me.

Thx, this is acceptable solution!
